File: sdk/typescript/src/observability/tracer.ts
import { trace, context, SpanStatusCode, SpanKind } from '@opentelemetry/api';

// AI Agent Semantic Conventions (matching Go implementation)
export const AI_AGENT_ATTRIBUTES = {
  // Agent identification
  AGENT_ID: 'ai.agent.id',
  AGENT_TYPE: 'ai.agent.type',
  AGENT_VERSION: 'ai.agent.version',
  
  // Request attributes
  REQUEST_INTENT: 'ai.request.intent',
  REQUEST_CONFIDENCE: 'ai.request.confidence',
  REQUEST_ACTING_FOR: 'ai.request.acting_for',
  REQUEST_SCOPE: 'ai.request.scope',
  
  // Policy evaluation
  POLICY_NAME: 'ai.policy.name',
  POLICY_VERSION: 'ai.policy.version',
  POLICY_RESULT: 'ai.policy.result',
  
  // Decision attributes
  DECISION_TYPE: 'ai.decision.type',
  DECISION_CONFIDENCE: 'ai.decision.confidence',
  DECISION_HUMAN_REVIEW: 'ai.decision.human_review',
  DECISION_RISK_SCORE: 'ai.decision.risk_score',
  DECISION_OUTCOME: 'ai.decision.outcome',
  
  // Multi-agent correlation
  SWARM_ID: 'ai.swarm.id',
  DELEGATION_CHAIN: 'ai.delegation.chain',
  PARENT_AGENT: 'ai.parent.agent',
  CHILD_AGENT: 'ai.child.agent',
  
  // Performance metrics
  LATENCY_MS: 'ai.latency.ms',
  CONFIDENCE_GATE_MS: 'ai.latency.confidence_gate_ms',
  POLICY_EVAL_MS: 'ai.latency.policy_eval_ms',
  RISK_EVAL_MS: 'ai.latency.risk_eval_ms',
} as const;

// Agent types
export const AGENT_TYPES = {
  AUTONOMOUS: 'autonomous',
  ASSISTED: 'assisted',
  HUMAN: 'human',
} as const;

// Decision types
export const DECISION_TYPES = {
  AUTONOMOUS: 'autonomous',
  HUMAN_REVIEW: 'human_review',
  DENIED: 'denied',
} as const;

export interface AgentSpanOptions {
  agentId: string;
  agentType?: string;
  action?: string | undefined;
  confidence?: number | undefined;
  actingFor?: string | undefined;
  scope?: string | undefined;
}

export interface DecisionMetrics {
  allowed: boolean;
  requiresHumanReview: boolean;
  confidence: number;
  riskScore: number;
  outcome: string;
}

export interface LatencyMetrics {
  totalMs: number;
  confidenceGateMs?: number;
  policyEvalMs?: number;
  riskEvalMs?: number;
}

/**
 * AgentTracer provides AI agent-aware tracing capabilities for the TypeScript SDK
 */
export class AgentTracer {
  private tracer = trace.getTracer('lelu-sdk-typescript', '0.2.0');

  /**
   * Start a new span with AI agent semantic conventions
   */
  startAgentSpan(operationName: string, options: AgentSpanOptions) {
    const span = this.tracer.startSpan(operationName, {
      kind: SpanKind.CLIENT,
      attributes: {
        [AI_AGENT_ATTRIBUTES.AGENT_ID]: options.agentId,
        [AI_AGENT_ATTRIBUTES.AGENT_TYPE]: options.agentType || AGENT_TYPES.AUTONOMOUS,
        ...(options.action && { [AI_AGENT_ATTRIBUTES.REQUEST_INTENT]: options.action }),
        ...(options.confidence !== undefined && { [AI_AGENT_ATTRIBUTES.REQUEST_CONFIDENCE]: options.confidence }),
        ...(options.actingFor && { [AI_AGENT_ATTRIBUTES.REQUEST_ACTING_FOR]: options.actingFor }),
        ...(options.scope && { [AI_AGENT_ATTRIBUTES.REQUEST_SCOPE]: options.scope }),
      },
    });

    return span;
  }

  /**
   * Start an authorization span
   */
  startAuthorizationSpan(agentId: string, action: string, confidence: number) {
    return this.startAgentSpan('ai.agent.authorize', {
      agentId,
      action,
      confidence,
      agentType: AGENT_TYPES.AUTONOMOUS,
    });
  }

  /**
   * Start a delegation span
   */
  startDelegationSpan(delegator: string, delegatee: string) {
    const span = this.startAgentSpan('ai.agent.delegate', {
      agentId: delegator,
    });

    span.setAttributes({
      [AI_AGENT_ATTRIBUTES.PARENT_AGENT]: delegator,
      [AI_AGENT_ATTRIBUTES.CHILD_AGENT]: delegatee,
      [AI_AGENT_ATTRIBUTES.DELEGATION_CHAIN]: `${delegator}→${delegatee}`,
    });

    return span;
  }

  /**
   * Start a swarm orchestration span
   */
  startSwarmSpan(swarmId: string, orchestrator: string, agents: string[]) {
    const span = this.startAgentSpan('ai.swarm.orchestration', {
      agentId: orchestrator,
      agentType: 'orchestrator',
    });

    span.setAttributes({
      [AI_AGENT_ATTRIBUTES.SWARM_ID]: swarmId,
      'ai.swarm.orchestrator': orchestrator,
      'ai.swarm.agents': agents,
      'ai.swarm.agent_count': agents.length,
    });

    return span;
  }

  /**
   * Record policy evaluation details
   */
  recordPolicyEvaluation(
    span: any,
    policyName: string,
    policyVersion: string,
    result: string,
    latencyMs: number
  ) {
    if (!span) return;

    span.setAttributes({
      [AI_AGENT_ATTRIBUTES.POLICY_NAME]: policyName,
      [AI_AGENT_ATTRIBUTES.POLICY_VERSION]: policyVersion,
      [AI_AGENT_ATTRIBUTES.POLICY_RESULT]: result,
      [AI_AGENT_ATTRIBUTES.POLICY_EVAL_MS]: latencyMs,
    });
  }

  /**
   * Record the final authorization decision
   */
  recordDecision(span: any, metrics: DecisionMetrics) {
    if (!span) return;

    const decisionType = metrics.requiresHumanReview
      ? DECISION_TYPES.HUMAN_REVIEW
      : metrics.allowed
      ? DECISION_TYPES.AUTONOMOUS
      : DECISION_TYPES.DENIED;

    span.setAttributes({
      [AI_AGENT_ATTRIBUTES.DECISION_TYPE]: decisionType,
      [AI_AGENT_ATTRIBUTES.DECISION_CONFIDENCE]: metrics.confidence,
      [AI_AGENT_ATTRIBUTES.DECISION_HUMAN_REVIEW]: metrics.requiresHumanReview,
      [AI_AGENT_ATTRIBUTES.DECISION_RISK_SCORE]: metrics.riskScore,
      [AI_AGENT_ATTRIBUTES.DECISION_OUTCOME]: metrics.outcome,
    });

    // Set span status based on decision
    if (!metrics.allowed && !metrics.requiresHumanReview) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: 'Authorization denied' });
    } else {
      span.setStatus({ code: SpanStatusCode.OK });
    }
  }

  /**
   * Record operation latency
   */
  recordLatency(span: any, metrics: LatencyMetrics) {
    if (!span) return;

    span.setAttributes({
      [AI_AGENT_ATTRIBUTES.LATENCY_MS]: metrics.totalMs,
      ...(metrics.confidenceGateMs && { [AI_AGENT_ATTRIBUTES.CONFIDENCE_GATE_MS]: metrics.confidenceGateMs }),
      ...(metrics.policyEvalMs && { [AI_AGENT_ATTRIBUTES.POLICY_EVAL_MS]: metrics.policyEvalMs }),
      ...(metrics.riskEvalMs && { [AI_AGENT_ATTRIBUTES.RISK_EVAL_MS]: metrics.riskEvalMs }),
    });
  }

  /**
   * Inject correlation context for multi-agent tracing
   */
  injectCorrelationContext(span: any, delegationChain: string) {
    if (!span) return;

    span.setAttributes({
      [AI_AGENT_ATTRIBUTES.DELEGATION_CHAIN]: delegationChain,
    });
  }

  /**
   * Execute a function within an agent span context
   */
  async withAgentSpan<T>(
    operationName: string,
    options: AgentSpanOptions,
    fn: (span: any) => Promise<T>
  ): Promise<T> {
    const span = this.startAgentSpan(operationName, options);
    
    try {
      const result = await context.with(trace.setSpan(context.active(), span), () => fn(span));
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (error) {
      span.recordException(error as Error);
      span.setStatus({ code: SpanStatusCode.ERROR, message: (error as Error).message });
      throw error;
    } finally {
      span.end();
    }
  }
}

// Export a default instance
export const agentTracer = new AgentTracer();

File: engine/internal/observability/tracer.go
// Package observability provides advanced observability capabilities for AI agents.
// This implements Phase 1 of the Advanced Observability & Telemetry roadmap.
package observability

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"
)

// AI Agent Semantic Conventions
// These follow OpenTelemetry semantic conventions for AI systems
const (
	// Agent identification
	AttrAgentID       = "ai.agent.id"
	AttrAgentType     = "ai.agent.type"
	AttrAgentVersion  = "ai.agent.version"
	
	// Request attributes
	AttrRequestIntent     = "ai.request.intent"
	AttrRequestConfidence = "ai.request.confidence"
	AttrRequestActingFor  = "ai.request.acting_for"
	AttrRequestScope      = "ai.request.scope"
	
	// Policy evaluation
	AttrPolicyName    = "ai.policy.name"
	AttrPolicyVersion = "ai.policy.version"
	AttrPolicyResult  = "ai.policy.result"
	
	// Decision attributes
	AttrDecisionType       = "ai.decision.type"
	AttrDecisionConfidence = "ai.decision.confidence"
	AttrDecisionHumanReview = "ai.decision.human_review"
	AttrDecisionRiskScore  = "ai.decision.risk_score"
	AttrDecisionOutcome    = "ai.decision.outcome"
	
	// Multi-agent correlation
	AttrSwarmID         = "ai.swarm.id"
	AttrDelegationChain = "ai.delegation.chain"
	AttrParentAgent     = "ai.parent.agent"
	AttrChildAgent      = "ai.child.agent"
	
	// Performance metrics
	AttrLatencyMs        = "ai.latency.ms"
	AttrConfidenceGateMs = "ai.latency.confidence_gate_ms"
	AttrPolicyEvalMs     = "ai.latency.policy_eval_ms"
	AttrRiskEvalMs       = "ai.latency.risk_eval_ms"
)

// Agent types
const (
	AgentTypeAutonomous = "autonomous"
	AgentTypeAssisted   = "assisted"
	AgentTypeHuman      = "human"
)

// Decision types
const (
	DecisionTypeAutonomous = "autonomous"
	DecisionTypeHumanReview = "human_review"
	DecisionTypeDenied     = "denied"
)

// AgentTracer provides AI agent-aware tracing capabilities
type AgentTracer struct {
	tracer trace.Tracer
}

// NewAgentTracer creates a new agent-aware tracer
func NewAgentTracer(serviceName string) *AgentTracer {
	return &AgentTracer{
		tracer: otel.Tracer(serviceName),
	}
}

// StartAgentSpan starts a new span with AI agent semantic conventions
func (t *AgentTracer) StartAgentSpan(ctx context.Context, operationName string, agentID string, attrs ...attribute.KeyValue) (context.Context, trace.Span) {
	// Add agent identification attributes
	agentAttrs := []attribute.KeyValue{
		attribute.String(AttrAgentID, agentID),
		attribute.String(AttrAgentType, AgentTypeAutonomous),
	}
	agentAttrs = append(agentAttrs, attrs...)
	
	ctx, span := t.tracer.Start(ctx, operationName)
	span.SetAttributes(agentAttrs...)
	
	return ctx, span
}

// StartAuthorizationSpan starts a span for agent authorization with full context
func (t *AgentTracer) StartAuthorizationSpan(ctx context.Context, agentID, action string, confidence float64) (context.Context, trace.Span) {
	return t.StartAgentSpan(ctx, "ai.agent.authorize",
		agentID,
		attribute.String(AttrRequestIntent, action),
		attribute.Float64(AttrRequestConfidence, confidence),
		attribute.String(AttrDecisionType, DecisionTypeAutonomous),
	)
}

// StartDelegationSpan starts a span for multi-agent delegation
func (t *AgentTracer) StartDelegationSpan(ctx context.Context, delegator, delegatee string) (context.Context, trace.Span) {
	return t.StartAgentSpan(ctx, "ai.agent.delegate",
		delegator,
		attribute.String(AttrParentAgent, delegator),
		attribute.String(AttrChildAgent, delegatee),
		attribute.String(AttrDelegationChain, fmt.Sprintf("%s→%s", delegator, delegatee)),
	)
}

// StartSwarmSpan starts a span for swarm orchestration
func (t *AgentTracer) StartSwarmSpan(ctx context.Context, swarmID string, orchestrator string) (context.Context, trace.Span) {
	return t.StartAgentSpan(ctx, "ai.swarm.orchestration",
		orchestrator,
		attribute.String(AttrSwarmID, swarmID),
		attribute.String(AttrAgentType, "orchestrator"),
	)
}

// RecordPolicyEvaluation records policy evaluation details in the span
func (t *AgentTracer) RecordPolicyEvaluation(span trace.Span, policyName, policyVersion, result string, latencyMs float64) {
	if span == nil {
		return
	}
	span.SetAttributes(
		attribute.String(AttrPolicyName, policyName),
		attribute.String(AttrPolicyVersion, policyVersion),
		attribute.String(AttrPolicyResult, result),
		attribute.Float64(AttrPolicyEvalMs, latencyMs),
	)
}

// RecordDecision records the final authorization decision
func (t *AgentTracer) RecordDecision(span trace.Span, allowed bool, requiresHumanReview bool, confidence, riskScore float64, outcome string) {
	if span == nil {
		return
	}
	
	decisionType := DecisionTypeAutonomous
	if requiresHumanReview {
		decisionType = DecisionTypeHumanReview
	} else if !allowed {
		decisionType = DecisionTypeDenied
	}
	
	span.SetAttributes(
		attribute.String(AttrDecisionType, decisionType),
		attribute.Float64(AttrDecisionConfidence, confidence),
		attribute.Bool(AttrDecisionHumanReview, requiresHumanReview),
		attribute.Float64(AttrDecisionRiskScore, riskScore),
		attribute.String(AttrDecisionOutcome, outcome),
	)
}

// RecordLatency records operation latency
func (t *AgentTracer) RecordLatency(span trace.Span, totalMs, confidenceGateMs, policyEvalMs, riskEvalMs float64) {
	if span == nil {
		return
	}
	span.SetAttributes(
		attribute.Float64(AttrLatencyMs, totalMs),
		attribute.Float64(AttrConfidenceGateMs, confidenceGateMs),
		attribute.Float64(AttrPolicyEvalMs, policyEvalMs),
		attribute.Float64(AttrRiskEvalMs, riskEvalMs),
	)
}

// InjectCorrelationContext injects correlation context for multi-agent tracing
func (t *AgentTracer) InjectCorrelationContext(ctx context.Context, parentSpan trace.Span, delegationChain string) context.Context {
	if parentSpan == nil {
		return ctx
	}
	
	// Add correlation attributes to the parent span
	parentSpan.SetAttributes(
		attribute.String(AttrDelegationChain, delegationChain),
	)
	
	return ctx
}

// ExtractCorrelationContext extracts correlation context from incoming requests
func (t *AgentTracer) ExtractCorrelationContext(ctx context.Context) (string, string) {
	span := trace.SpanFromContext(ctx)
	if span == nil {
		return "", ""
	}
	
	// In a real implementation, you would extract these from span context
	// For now, we'll return empty strings as placeholders
	return "", ""
}